### PR TITLE
ARTEMIS-4355 Update Curator to 5.5.0; Zookeeper 3.8.2

### PR DESCRIPTION
**Pull request by amarkevich**

### CODE REVIEW

**Comment by jbertram** [comment]
Why were the test changes necessary?

**Comment by amarkevich** [comment]
There were compilation errors because `getQuorumPeer` removed

File: artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java [hunk 1]
```diff
@@ -84,8 +82,6 @@ public void setupEnv() throws Throwable {
       }
       testingServer = new TestingCluster(clusterSpecs);
       testingServer.start();
-      // start waits for quorumPeer!=null but not that it has started...
-      Wait.waitFor(this::ensembleHasLeader);
```
Comment (gemmellr) [inline artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java:88]: Did this actually change on the Curator side, or did you just remove the 'wait for start' as the getQuorumPeer method went away?

File: artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java [hunk 2]
```diff
@@ -337,21 +333,8 @@ public void beNotifiedOfAlreadyUnavailableManagerAfterAddingListener() throws Ex
...
-         return true;
-      }
-      long leaderId = server.getQuorumPeer().getLeaderId();
-      long id = server.getQuorumPeer().getId();
-      return id == leaderId;
-   }
-
    private void stopMajorityNotLeaderNodes(boolean fromLast) throws Exception {
-      List<TestingZooKeeperServer> followers = testingServer.getServers().stream().filter(Predicates.not(CuratorDistributedLockTest::isLeader)).collect(Collectors.toList());
+      List<TestingZooKeeperServer> followers = testingServer.getServers();
```
Comment (gemmellr) [inline artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java:337]: Is this now just stopping 'a majority of nodes' rather than the original 'majority of non-leader nodes'?

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java [hunk 3]
```diff
@@ -48,15 +47,7 @@ public ZookeeperPluggableQuorumPeerTest() {
...
       backup = new BrokerControl("primary-peer-b", JMX_PORT_BACKUP, "zkReplicationPrimaryPeerB", BACKUP_PORT_OFFSET);
-      brokers = new LinkedList(Arrays.asList(primary, backup));
-   }
-
-   @Test
-   @Override
-   public void testBackupFailoverAndPrimaryFailback() throws Exception {
-      // peers don't request fail back by default
-      // just wait for setup to avoid partial stop of zk via fast tear down with async setup
-      Wait.waitFor(this::ensembleHasLeader);
```
Comment (gemmellr) [inline tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java:59]: Again, did something change on the Curator side, or is this now potentially doing the very 'partial stop of zk via fast tear down with async setup' this was specifically there to avoid?

File: artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java (hunk 1 above)
Comment (amarkevich) [reply artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java:88]: `ensembleHasLeader` used 'getQuorumPeer' inside

File: artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java (hunk 2 above)
Comment (amarkevich) [reply artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java:337]: renamed

File: artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java (hunk 1 above)
Comment (gemmellr) [reply artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java:88]: I realise, which is why I mentioned it in my question. What I wondered is whether the reason it did so, to wait for _actual startup_, has changed on the curator side. If it didn't then you have just removed a step that I expect was there for a reason.

File: artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java (hunk 2 above)
Comment (gemmellr) [reply artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java:337]: I was not really pointing out the name wasnt descriptive, but that you appeared to have changed _what the method is trying to do_. As that could be a difference that alters expected behaviour for tests using it, it isnt immediately obvious as being desirable.

**Comment by amarkevich** [comment]
`[INFO] Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 188.484 s - in org.apache.activemq.artemis.tests.smoke.quorum.ZookeeperPluggableQuorumPeerTest
`

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java [hunk 4]
```diff
@@ -85,9 +85,9 @@ public static Collection getParameters() {
          {"SunJCE", "JCEKS", false, false},
          {"SUN", "JKS", false, false},
          {"SunJSSE", "PKCS12", false, false},
-         {"JCEKS", null, true, false}, // for compatibility with old keyStoreProvider
-         {"JKS", null, true, false},   // for compatibility with old keyStoreProvider
-         {"PKCS12", null, true, false} // for compatibility with old keyStoreProvider
+         {"JCEKS", null, false, false}, // for compatibility with old keyStoreProvider
+         {"JKS", null, false, false},   // for compatibility with old keyStoreProvider
+         {"PKCS12", null, false, false} // for compatibility with old keyStoreProvider
```
Comment (jbertram) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java:90]: I don't understand why these changes were necessary. This eliminates the tests for the warning log message.

Comment (amarkevich) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java:90]: To pass tests for this PR - otherwise quorum tests are not running. Expected changes in first commit only

File: artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java (hunk 1 above)
Comment (amarkevich) [reply artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java:87]: reverted

File: artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java (hunk 2 above)
Comment (amarkevich) [reply artemis-quorum-ri/src/test/java/org/apache/activemq/artemis/quorum/zookeeper/CuratorDistributedLockTest.java:336]: Since leader api was removed there is no significant choice which instance stopped

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java (hunk 3 above)
Comment (amarkevich) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java]: reverted with different check. In this PR quorum tests are enabled to show that its passed

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/common/SmokeTestBase.java [hunk 5]
```diff
@@ -158,24 +156,6 @@ protected static final void recreateBrokerDirectory(final String homeInstance) {
...
-            System.out.println("awaiting server " + uri + " start at ");
-            Thread.sleep(500);
-            continue;
-         }
-         return true;
-      }
-
-      return false;
-   }
-
```
Comment (jbertram) [inline tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/common/SmokeTestBase.java:178]: Is this change actually related to the ZooKeep & Curator upgrade? It looks like an unrelated removal of an unused method.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java (hunk 4 above)
Comment (amarkevich) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java]: this test started to fail because Zookeeper 3.8.x uses logback. rollback to 3.7.1

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/common/SmokeTestBase.java (hunk 5 above)
Comment (amarkevich) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/common/SmokeTestBase.java:178]: looks like actual logic is moved to ServerUtil and this method confuses when looking for corresponding check

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java (hunk 4 above)
Comment (amarkevich) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java]: I'm not sure that Zookeeper will work after such changes

Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java]: I took a look at the ZooKeeper code. As far as I can tell it's using SLF4J as the API (just like the broker), and Logback is the implementation. I think you'd be fine to exclude Logback via the `pom.xml`.

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java [hunk 6]
```diff
@@ -48,15 +47,15 @@ public ZookeeperPluggableQuorumPeerTest() {
...
+      brokers = Arrays.asList(primary, backup);
    }
 
    @Test
    @Override
    public void testBackupFailoverAndPrimaryFailback() throws Exception {
       // peers don't request fail back by default
       // just wait for setup to avoid partial stop of zk via fast tear down with async setup
-      Wait.waitFor(this::ensembleHasLeader);
+      Assert.assertTrue(awaitAsyncSetupCompleted(TimeUnit.SECONDS.toMillis(30L), TimeUnit.MILLISECONDS));
```
Comment (gemmellr) [inline tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java:58]: Being curious how you were able to remove the methods in the other class without affecting anything else, I had a look, and it appears like this different check will in fact do nothing really, as the called method just returns true. So its asserting true on something that just returns true.

The fact awaitAsyncSetupCompleted simply returns true, is interesting...though the effect of this change is still to no wait for startup, which it did before.

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/common/SmokeTestBase.java (hunk 5 above)
Comment (gemmellr) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/common/SmokeTestBase.java:178]: Thats fine, but Justin's correct point is it shouldn't be in the same commit as it isnt related.

Comment (amarkevich) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/common/SmokeTestBase.java]: moved to additional PR

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java (hunk 4 above)
Comment (amarkevich) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/ssl/CoreClientOverOneWaySSLTest.java]: done

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java (hunk 6 above)
Comment (amarkevich) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java:58]: Should I remove this dummy check everywhere?

**Comment by gemmellr** [comment]
FWIW I ran the earlier version of this (cd5887b0e8) through the full test suite in another CI env and it rather fell over in a heap, failing various tests (even though in various cases they seem to have done what was expected), and timing out and being terminated overall. I didnt bother to re-run it (in case it was a sporadic env issue) since the PR had changed singnificantly by that point, and has now again.

I ran c64e2e31a6 through the full suite today though and it seemed ok, even if I do still question some of the actual behaviour changes being made.

File: tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java (hunk 6 above)
Comment (amarkevich) [reply tests/smoke-tests/src/test/java/org/apache/activemq/artemis/tests/smoke/quorum/ZookeeperPluggableQuorumPeerTest.java]: just @Ignore'd

**Comment by gemmellr** [comment]
Still not that sure of some of the changes, but they dont seem to break the tests, and it is past time these were updated, so...in they go.
